### PR TITLE
TINY-6870: BDD tests for visualchars

### PR DESCRIPTION
**Pull request by JamesToohey**
Related Ticket: TINY-6870

Description of Changes:
BDD-style tests for `visualchars`

Pre-checks:
* [x] ~~Changelog entry added~~
* [x] ~~Tests have been added (if applicable)~~
* [x] ~~Branch prefixed with `feature/` for new features (if applicable)~~
* [x] ~~License headers added on new files (if applicable)~~

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/src/plugins/visualchars/test/ts/atomic/DataTest.ts [hunk 1]
```diff
@@ -1,22 +1,18 @@
...
 import * as Data from 'tinymce/plugins/visualchars/core/Data';
 
-UnitTest.test('atomic.tinymce.plugins.visualchars.DataTest', () => {
-  Assert.eq(
-    'should return correct selector',
-    'span.mce-a,span.mce-b',
-    Data.charMapToSelector({ a: 'a', b: 'b' })
-  );
+describe('atomic.tinymce.plugins.visualchars.DataTest', () => {
+  it('should return correct selector', () => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/visualchars/test/ts/atomic/DataTest.ts:7]: Does it make sense to break this up into 3 different tests? It wasn't what I was expecting, as I would have just made it 2 tests for each function with 3 assertions. eg:
```
it('charMapToSelector', () => {
  assert.equal('span.mce-a,span.mce-b', Data.charMapToSelector({ a: 'a', b: 'b' }), 'should return correct selector');
});

it('charMapToRegExp', () => {
  assert.equal('/[ab]/', Data.charMapToRegExp({ a: 'a', b: 'b' }).toString(), 'should return correct regexp');
  assert.equal('/[ab]/g', Data.charMapToRegExp({ a: 'a', b: 'b' }, true).toString(), 'should return correct global regexp');
});
```

File: modules/tinymce/src/plugins/visualchars/test/ts/atomic/HtmlTest.ts [hunk 2]
```diff
@@ -1,17 +1,21 @@
...
 import * as Html from 'tinymce/plugins/visualchars/core/Html';
 
-UnitTest.test('atomic.tinymce.plugins.visualchars.HtmlTest', () => {
-  Assert.eq(
-    'should return correct span',
-    '<span data-mce-bogus="1" class="mce-nbsp">' + Unicode.nbsp + '</span>',
-    Html.wrapCharWithSpan(Unicode.nbsp)
-  );
+describe('atomic.tinymce.plugins.visualchars.HtmlTest', () => {
+  it('should return correct span with nbsp', () => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/visualchars/test/ts/atomic/HtmlTest.ts:8]: Likewise, I would have just expected the one test here, as this is testing the module not `wrapCharWithSpan`. I guess the alternative option would be to wrap it inside a `context` call to highlight that's what it's testing.

File: modules/tinymce/src/plugins/visualchars/test/ts/browser/PluginTest.ts [hunk 3]
```diff
@@ -1,51 +1,53 @@
-import { ApproxStructure, Assertions, Log, Pipeline } from '@ephox/agar';
-import { UnitTest } from '@ephox/bedrock-client';
+// import { ApproxStructure, Assertions, Log, Pipeline } from '@ephox/agar';
```
Comment (ashfordneil) [inline modules/tinymce/src/plugins/visualchars/test/ts/browser/PluginTest.ts:1]: Do we still need this? Also reminder for that lint rule discussion from this morning.

**Review by ashfordneil — APPROVED** [review]

File: modules/tinymce/src/plugins/visualchars/test/ts/atomic/HtmlTest.ts (hunk 2 above)
Comment (lnewson) [reply modules/tinymce/src/plugins/visualchars/test/ts/atomic/HtmlTest.ts:8]: Also worth noting most current unit tests, test the function generally not one specific aspect of it (eg `DomUtilsTest` is a good example).

File: modules/tinymce/src/plugins/visualchars/test/ts/browser/PluginTest.ts (hunk 3 above)
Comment (JamesToohey) [reply modules/tinymce/src/plugins/visualchars/test/ts/browser/PluginTest.ts:1]: Sorry I missed that

File: modules/tinymce/src/plugins/visualchars/test/ts/atomic/DataTest.ts (hunk 1 above)
Comment (lnewson) [reply modules/tinymce/src/plugins/visualchars/test/ts/atomic/DataTest.ts:7]: As discussed, we're going to keep this style. However, this does need to be wrapped in a context so it's clear what part of `DataTest` it's testing.

- [ ] Wrap in context for each function being tested

File: modules/tinymce/src/plugins/visualchars/test/ts/atomic/HtmlTest.ts (hunk 2 above)
Comment (lnewson) [reply modules/tinymce/src/plugins/visualchars/test/ts/atomic/HtmlTest.ts:8]: This just needs to be wrapped in a context block now given our discussion at standup.

- [ ] Wrap in context

File: modules/tinymce/src/plugins/visualchars/test/ts/atomic/NodesTest.ts [hunk 4]
```diff
@@ -1,47 +1,39 @@
...
 import { Unicode } from '@ephox/katamari';
 import { SugarElement } from '@ephox/sugar';
+import { assert } from 'chai';
+
 import * as Nodes from 'tinymce/plugins/visualchars/core/Nodes';
 
-UnitTest.test('atomic.tinymce.plugins.visualchars.NodesTest', () => {
-  const testReplaceWithSpans = () => {
+describe('atomic.tinymce.plugins.visualchars.NodesTest', () => {
+  it('replace with spans', () => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/visualchars/test/ts/atomic/NodesTest.ts:10]: - [ ] Wrap in context also?

File: modules/tinymce/src/plugins/visualchars/test/ts/browser/DefaultStateTest.ts [hunk 5]
```diff
@@ -1,35 +1,31 @@
...
     plugins: 'visualchars',
     toolbar: 'visualchars',
     base_url: '/project/tinymce/js/tinymce',
     visualchars_default_state: true
-  }, success, failure);
+  }, [ Plugin, Theme ]);
+
+  it('tests the default visualchars state', async () => {
+    const editor = hook.editor();
+    editor.setContent('<p>a&nbsp;&nbsp;b</p>', { format: 'raw' });
```
Comment (lnewson) [inline modules/tinymce/src/plugins/visualchars/test/ts/browser/DefaultStateTest.ts:21]: Why is this using `raw`? as it didn't previously.

**Review by lnewson — APPROVED** [review]

**Pull request merged**